### PR TITLE
Fix 14 15 18

### LINKED ISSUE #14: Login API and Service
Create the login endpoint in the API and create the login service. 

A username and password should be validated and then if a valid combination, should be exchanged for a token. 

Token should be stored in a table which has the userid and token.

### LINKED ISSUE #15: Signup API and Service
Create the Signup Endpoint in the API and then create the signup service which will store the user information in the users database.

### LINKED ISSUE #18: Create an auth endpoint
This should validated if the token is currently valid and if it is invalid then return an unauthorized error.

### PR DESCRIPTION
**Pull request by enclark**
closes #14 
closes #15
closes #18 

Authentication

Added passwords and the ability to authenticate a user using a auth token

### CODE REVIEW

**Comment by jackcmeyer** [comment]
* In the future, use `git rebase master` to "merge" master to your branch. This will get rid of the ugly merge commits. 

* Also, if you notice that you have a lint error in your code. Use `git add <file_lint_error_is_in>`, `git commit --amend --no-edit`, `git push origin <branch_name> -f` to amend your previous commit to eliminate the lint error commits.

File: endor/src/error/InvalidCredentialsException.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { getActiveLogger } from '../utils/winston';
+
+const log = getActiveLogger();
+
+export default class InvalidCredentialsException {
+  constructor(message) {
+    this.message = message;
+    this.type = 'Invalid Credentials';
+    this.status = 400; // TODO I don't know about this
```
Comment (jackcmeyer) [inline endor/src/error/InvalidCredentialsException.js]: 401 would probably be a good error code here. 

https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses

File: endor/src/error/TokenNotFoundException.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import NotFoundException from './NotFoundException';
+
+export default class TokenNotFoundException extends NotFoundException {
+  constructor(message) {
+    super(message, 'Not Found', 404);
```
Comment (jackcmeyer) [inline endor/src/error/TokenNotFoundException.js]: I could be wrong here. But a TokenNotFoundException would mean that the token is not valid, meaning the user on the request would not be authenticated. Should this be a 401?

File: endor/src/index.js [hunk 3]
```diff
@@ -1,39 +1,57 @@
...
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(cookieParser());
 
+// TODO - reexamine if this is necessary for oauth2orize
+app.use(session({
+  secret: 'Super Secret Session Key',
+  saveUninitialized: true,
+  resave: true
+}));
+
```
Comment (jackcmeyer) [inline endor/src/index.js]: Did we reexamine? We should do this before merging. If it's done, please remove TODO

File: endor/src/routes/client.routes.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ *   ]
+ * }
+ */
+router.get('/clients/:userId', authController.isAuthenticated, clientController.getAllClients);
+
+export function setDependencies(newUserService, newClientService, newAuthService) {
+  clientService = newClientService;
+  clientController.setDependencies(clientService);
+  authController.setDependencies(newUserService, newClientService, newAuthService);
+}
```
Comment (jackcmeyer) [inline endor/src/routes/client.routes.js]: Follow API documentation for other routes so we can use apidoc for generation. 

Check out the README for how to generate docs.

File: endor/test/sequalize-mock.js [hunk 5]
```diff
@@ -1,29 +1,57 @@
 import Sequelize from 'sequelize';
 
 // eslint-disable-next-line import/no-unresolved
+import dbConfig from '../../dbConfig.json';
 
 // Added for convenience
 // eslint-disable-next-line prefer-destructuring
 const STRING = Sequelize.DataTypes.STRING;
+// eslint-disable-next-line prefer-destructuring
+const BOOLEAN = Sequelize.DataTypes.BOOLEAN;
```
[the remark this answers is not in the document]
Comment (nischayv) [reply endor/test/sequalize-mock.js]: @jackcmeyer @enclark Can you not just do const {STRING, BOOLEAN } = Sequelize.DataTypes?

[the remark this answers is not in the document]
Comment (enclark) [reply endor/test/sequalize-mock.js]: I changed it to use destructuring like @nischayv suggested

**Comment by nischayv** [comment]
Might want to run `npm run lint` in the future before each commit :)

**Comment by jackcmeyer** [comment]
@nischayv if I'm feeling ambitious, I'm going to make a pre-commit hook that can be distributed which will run npm lint before :)

File: endor/README.md [hunk 6]
```diff
@@ -90,3 +90,41 @@ sequelize.User.findAll({
...
+###Basic-Auth
+Uses the user's username and password in the Authentication header to authenticate
+the user.
+
+###Client-Basic
+Uses the client's clientId and secret like a username and password to authenticate and
+to protect a user's credentials.
+
+###Bearer
+A token is used to authenticate the user.
```
Comment (jackcmeyer) [inline endor/README.md]: Your headers are not coming out properly in the README.md file, there needs to be a space between the # group and the heading. For example 

`# Authentication` creates
# Authentication

**Comment by nischayv** [comment]
Gonna merge this!!

**Pull request merged**